### ISSUE TITLE
DPI Scaling Issues

### ISSUE THREAD

**Issue by HACKERALERT** [AllenDang/giu#350]
Hi,

There are multiple DPI scaling issues I believe. For example, the image on the left is a normal display, and the one on the right is a high DPI (factor 2) display. Notice that there's a larger gap in the high DPI display screenshot.

![Screenshot (183)](https://user-images.githubusercontent.com/48808396/134821143-9ee9f474-c2f3-4c91-a4b4-c6656853d2ca.png)

In addition, I have to use multiple hacks to get the scaling correct. For example, I have to do divide by `dpi` to get the correct scaling on some types of widgets, otherwise, on the high DPI display, it will be too big:
```go
giu.InputText(&password).Flags(passwordState).Size(302/dpi)
```
```go
giu.Button(s("Change")).Size(bw/dpi, 0).OnClick(func() {
```
```go
giu.Dummy(size/dpi, 0).Build()
```
```go
giu.ProgressBar(shredProgress).Overlay(shredOverlay).Size(size/dpi, 0),
```

There's even a line where I had to divide twice to get things to work:
```go
giu.InputText(&outputFile).Size(dw / dpi / dpi).Flags(giu.InputTextFlagsReadOnly),
```

It might be a good idea to fix this so that giu looks consistent on all displays without having to divide by `dpi`. I'm not sure if this is an Imgui limitation, but it's worth a look into because it will make development with giu a lot easier! For reference, my entire UI code is here: https://github.com/HACKERALERT/Picocrypt/blob/main/src/unstable/Picocrypt.go#L225. Thank you!

**Comment by AllenDang** [comment]
@HACKERALERT The root cause is at the time giu invent it's own DPI scaling mechanism, imgui doesn't have one. And for recent version of imgui, it comes with a inside DPI scaling, that causes the size is scaled twice...
I may need more time to fix this as you know, a national festivel is coming.

**Comment by AllenDang** [comment]
@HACKERALERT Try again with newest code. I removed all dpi scaling from giu side.

**Comment by HACKERALERT** [comment]
Not quite. Perhaps ImGui is having issues?
Normal DPI: ![Screenshot (184)](https://user-images.githubusercontent.com/48808396/134976623-18997bbc-6529-4f84-9884-b7563ecc75f0.png)
High DPI: ![Screenshot (196)](https://user-images.githubusercontent.com/48808396/134976633-750dff62-82aa-45bb-bcd2-01016d122b26.png)

**Comment by AllenDang** [comment]
@macbutch Pls help me to do a test to see whether DPI scaling works. Many thanks!

**Comment by HACKERALERT** [comment]
@AllenDang just gave the master branch a try, it still has the same issue and looks the same as my second picture.

**Comment by macbutch** [comment]
@AllenDang just did a quick test - the size of the window is correct but everything inside is tiny. Happy to do more testing as it evolves but this latest build doesn't seem to scale for me. Is there something I need to add?

**Comment by HACKERALERT** [comment]
@AllenDang Almost there! Only a few more issues to sort out 😄 .
![Screenshot (197)](https://user-images.githubusercontent.com/48808396/137603742-4bbc7946-19aa-4f19-ac84-ce725a8d33d2.png)

**Comment by macbutch** [comment]
OK. This latest one is much better than before but I have a few issues with the master branch right now:

1. some spaces and sizes come out differently on High DPI: an image/icon is about half the size and in another case some buttons overflow. It's very close otherwise but there is less padding of text inside buttons on Windows vs MacOS for me.
2. There seems to be some change in behaviour for Popups which I've not quite understood and one of my Popups closes unexpectedly sometimes.

I could likely live with (or workaround) the first issue but the second is quite annoying (popup closing while still being used).  Has anything changed there?

**Comment by macbutch** [comment]
@AllenDang A tiny bit more feedback but I'm away from my Windows computer for a few days and can't test more until the weekend. It's much improved but there are a few cases where things aren't quite the right size. So where I've got some careful calculations of size etc to make sure everything fits on MacOS those things don't quite come out the same on Windows (so I end up with scrolls bars where I don't want them, etc). I've also noticed that images come out smaller (this isn't a big deal for my app but I think that Image still needs to be scaled up somehow). I'll report more when I can.
